### ISSUE TITLE
Is there any demo code about how to use apex for gradient accumulation?

### ISSUE THREAD

**Issue by StillKeepTry** [NVIDIA/apex#374]
Thanks for your works. Is there any demo code about how to use apex for gradient accumulation?

**Comment by ptrblck** [comment]
Hi @StillKeepTry,

[this code snippet](https://nvidia.github.io/apex/advanced.html#gradient-accumulation-across-iterations) should work for gradient accumulation.

**Comment by mcarilli** [comment]
If you are trying to use gradient accumulation to simulate a larger batch size, you should also divide the loss by the number of accumulation steps:
```
    loss = loss/iters_to_accumulate

    if iter%iters_to_accumulate == 0:
        # Every iters_to_accumulate iterations, unscale and step
        with amp.scale_loss(loss, optimizer) as scaled_loss:
            scaled_loss.backward()
        # Gradient clipping if desired:
        # torch.nn.utils.clip_grad_norm_(amp.master_params(optimizer), max_norm)
        optimizer.step()
        optimizer.zero_grad()
    else:
        # Otherwise, accumulate gradients, don't unscale or step.
        with amp.scale_loss(loss, optimizer) as scaled_loss:
            scaled_loss.backward()
```

I've [updated the documentation](https://nvidia.github.io/apex/advanced.html#gradient-accumulation-across-iterations) accordingly.

**Comment by StillKeepTry** [comment]
Thank you for your response.

**Comment by unendin** [comment]
> If you are trying to use gradient accumulation to simulate a larger batch size, you should also divide the loss by the number of accumulation steps:
> 
> ```
>     loss = loss/iters_to_accumulate
> 
>     if iter%iters_to_accumulate == 0:
>         # Every iters_to_accumulate iterations, unscale and step
>         with amp.scale_loss(loss, optimizer) as scaled_loss:
>             scaled_loss.backward()
>         # Gradient clipping if desired:
>         # torch.nn.utils.clip_grad_norm_(amp.master_params(optimizer), max_norm)
>         optimizer.step()
>         optimizer.zero_grad()
>     else:
>         # Otherwise, accumulate gradients, don't unscale or step.
>         with amp.scale_loss(loss, optimizer) as scaled_loss:
>             scaled_loss.backward()
> ```
> 
> I've [updated the documentation](https://nvidia.github.io/apex/advanced.html#gradient-accumulation-across-iterations) accordingly.



> If you are trying to use gradient accumulation to simulate a larger batch size, you should also divide the loss by the number of accumulation steps:
> 
> ```
>     loss = loss/iters_to_accumulate
> 
>     if iter%iters_to_accumulate == 0:
>         # Every iters_to_accumulate iterations, unscale and step
>         with amp.scale_loss(loss, optimizer) as scaled_loss:
>             scaled_loss.backward()
>         # Gradient clipping if desired:
>         # torch.nn.utils.clip_grad_norm_(amp.master_params(optimizer), max_norm)
>         optimizer.step()
>         optimizer.zero_grad()
>     else:
>         # Otherwise, accumulate gradients, don't unscale or step.
>         with amp.scale_loss(loss, optimizer) as scaled_loss:
>             scaled_loss.backward()
> ```
> 
> I've [updated the documentation](https://nvidia.github.io/apex/advanced.html#gradient-accumulation-across-iterations) accordingly.

The two calls to amp are identical, though the comments and conditional suggest otherwise. Am I missing something?

**Comment by mcarilli** [comment]
Yes, they are identical in both branches.  The `don't unscale` part of the comment `Otherwise, accumulate gradients, don't unscale or step.` is actually incorrect.  Gradients are unscaled as they are accumulated in both branches.  

[Updated the example snippet again](https://nvidia.github.io/apex/advanced.html#gradient-accumulation-across-iterations) to emphasize that the backward pass does not need to change between accumulation iterations and stepping iterations.

**Comment by chenyangyu1988** [comment]
@mcarilli How does it work in gradients accumulation if there is overflow in one of the iterations? I think there might be some issue because we are doing unscaling param by param, and it looks like that it will all if any single iteration have overflow, is it correct?

**Comment by mcarilli** [comment]
With Amp+gradient accumulation, (unscaled) gradients are accumulated every backward pass, just as they would be without Amp.  Also, gradients are checked for infs/nans every backward pass.  Therefore, if we are accumulating over `N` iterations, if infs/nans are produced in any of those `N` backward passes, the resulting accumulated gradients will contain infs/nans, and the `optimizer.step()` for that entire batch of `N` accumulations will be skipped.

In general, with or without gradient accumulation, the decision to skip a given `step()` is made globally:  If _any_ of the params has an inf/nan gradient, the step will be skipped for _all_ params.  I don't think it makes mathematical sense to update only the subset of params that happen not to have inf/nan gradients.

**Comment by abhinavs95** [comment]
@mcarilli In the example snippet you provided for gradient accumulation, is communication (all reduce) overlapped with the backward pass?

**Comment by krishansubudhi** [comment]
@mcarilli how can i disable all reduce in the else case? It does not make sense to enable all reduce at every step.

**Comment by ggaemo** [comment]
```
# If your intent is to simulate a larger batch size using gradient accumulation,
# you can divide the loss by the number of accumulation iterations (so that gradients
# will be averaged over that many iterations):
loss = loss/iters_to_accumulate

with amp.scale_loss(loss, optimizer) as scaled_loss:
    scaled_loss.backward()

# Every iters_to_accumulate iterations, call step() and reset gradients:
if iter%iters_to_accumulate == 0:
    # Gradient clipping if desired:
    # torch.nn.utils.clip_grad_norm_(amp.master_params(optimizer), max_norm)
    optimizer.step()
    optimizer.zero_grad()

```

In the snippet above, there is loss / args.iters_to_accumulate
```
if iter%iters_to_accumulate == 0:
    # Every iters_to_accumulate iterations, unscale and step
    with amp.scale_loss(loss, optimizer) as scaled_loss:
        scaled_loss.backward()
    optimizer.step()
    optimizer.zero_grad()
else:
    # Otherwise, accumulate gradients, don't unscale or step.
    with amp.scale_loss(loss, optimizer, delay_unscale=True) as scaled_loss:
        scaled_loss.backward()
```

in the snippet below, there isn't.

Should there be a loss / args.iters_to_accumulate in the snippet below as also?

**Comment by shenfe** [comment]
Is there any solution like [torch.nn.parallel.DistributedDataParallel.no_sync](https://pytorch.org/docs/stable/generated/torch.nn.parallel.DistributedDataParallel.html#torch.nn.parallel.DistributedDataParallel.no_sync) when it does not need sync if `iter%iters_to_accumulate != 0` ?
